### PR TITLE
chips/nrf5x: disallow registering multiple interrupts for a single GPIO

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This pull request ensures that only a single interrupt can be configured for a GPIO pin on the nRF5x at any given time.

The current nRF5x GPIO implementation allows registering multiple interrupts into the interrupt channels even for a single GPIO. However, all of those interrupts will still be routed to the same `GPIOPin`'s client. Also, a second enabled interrupt would not be disabled by a single call to `disable_interrupt`.

Apart from these issues, other Tock code is not written under the assumption that there can be multiple interrupts fired by a single Pin event. For instance, when restarting an app, this will cause the button driver to issue two callbacks for every actual physical event to the restarted app.

This PR extends `GPIOPin` with an additional `allocated_channel` field that keeps track of the channel allocated. Only one interrupt channel can be allocated per GPIO. Using `allocated_channel` also avoids scanning over all channels in some cases.

### Testing Strategy

This pull request was tested by restarting an app and ensuring that only a single button interrupt is delivered.

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: chips/nrf5x/src/gpio.rs [hunk 1]
```diff
@@ -504,26 +506,41 @@ impl<'a> hil::gpio::Interrupt<'a> for GPIOPin<'a> {
...
+        self.gpiote_registers.intenset.set(1 << channel);
     }
 
     fn disable_interrupts(&self) {
-        if let Ok(channel) = self.find_channel(self.pin) {
+        if let Some(channel) = self.allocated_channel.get() {
             self.gpiote_registers.config[channel]
                 .write(Config::MODE::CLEAR + Config::PSEL::CLEAR + Config::POLARITY::CLEAR);
             self.gpiote_registers.intenclr.set(1 << channel);
+	    self.allocated_channel.take();
```
Comment (brghena) [inline chips/nrf5x/src/gpio.rs:543]: Change to `.clear()` to make the intent more obvious?

File: chips/nrf5x/src/gpio.rs [hunk 2]
```diff
@@ -504,26 +506,41 @@ impl<'a> hil::gpio::Interrupt<'a> for GPIOPin<'a> {
...
-            let pin: u32 = (GPIO_PER_PORT as u32 * self.port as u32) + self.pin as u32;
-            self.gpiote_registers.config[channel]
-                .write(Config::MODE::Event + Config::PSEL.val(pin) + polarity);
-            self.gpiote_registers.intenset.set(1 << channel);
+        let channel = if let Some(chan) = self.allocated_channel.get() {
+            // We only support one interrupt mode per pin, despite the
+            // hardware supporting multiple. This is to comply with
+            // expectations in other Tock components, such as the button
+            // driver which re-registers interrupts for a restarted app,
+            // assuming the old ones to be overwritten.
```
Comment (brghena) [inline chips/nrf5x/src/gpio.rs:514]: This change seems reasonable, but isn't the bigger issue that an app 1) made a request that allocated some resource and then 2) stopped (or restarted in this case) but didn't free the resource. So if an app stops you'll keep getting GPIO interrupts for it, even though there's no client?

I wonder if we handle this resource dropped issue _anywhere_ in Tock.

**Review by brghena — APPROVED** [review]
Small change and a question, but I do approve

File: chips/nrf5x/src/gpio.rs (hunk 1 above)
Comment (lschuermann) [reply chips/nrf5x/src/gpio.rs]: Changed and fixed formatting.

**Review by brghena — APPROVED** [review]

File: chips/nrf5x/src/gpio.rs (hunk 2 above)
Comment (lschuermann) [reply chips/nrf5x/src/gpio.rs:514]: Yes, this is an issue and I believe it to be an issue at large in the Tock ecosystem. We don't currently have a good way of testing or exercising these cases, and the fact that an app restart can trigger this relatively simple bug seems to be a symptom of that.

However, for this particular case, it's actually quite clear what we're ought to do:
https://github.com/tock/tock/blob/ae86c8366b064391571680c4f49508065a3d3569/capsules/core/src/button.rs#L245-L250

The driver itself is managing these subscriptions and multiplexing multiple applications onto interrupts. As such, when an application dies, there might actually be other apps that are still interested in this subscription. The driver does then exercise the above logic to deregister clients when they're no longer needed, lazily upon getting a callback.

The issue here is really just that we assume `enable_interrupt` to be an idempotent function, when it is not for the nRF. And that actually seems like a reasonable design to me, if only for this subsystem.

**Pull request merged**